geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java

```diff
@@ -203,49 +279,122 @@ public int hstrlen(ByteArrayWrapper field) {
+        resultList);
+  }
+
+  private void removeHSCANSnapshot(UUID clientID) {
+    this.hScanSnapShots.remove(clientID);
+    this.hScanSnapShotCreationTimes.remove(clientID);
+
+    if (this.hScanSnapShots.isEmpty()) {
+      shutDownHscanSnapshotScheduledRemoval();
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private Pair<Integer, List<Object>> getResultsPair(List<ByteArrayWrapper> keysSnapShot,
+      int startCursor,
+      int count,
+      Pattern matchPattern) {
+
+    int indexOfKeys = startCursor;
+
+    List<ByteArrayWrapper> resultList = new ArrayList<>();
+
+    for (int index = startCursor; index < keysSnapShot.size(); index++) {
+
+      if ((index - startCursor) == count) {
+        break;
+      }
+
+      ByteArrayWrapper key = keysSnapShot.get(index);
+      indexOfKeys++;
+
+      ByteArrayWrapper value = hash.get(key);
+      if (value == null) {
         continue;
       }
 
       if (matchPattern != null) {
         if (matchPattern.matcher(key.toString()).matches()) {
-          returnList.add(key);
-          returnList.add(value);
-          numElements++;
+          resultList.add(key);
+          resultList.add(value);
         }
       } else {
-        returnList.add(key);
-        returnList.add(value);
-        numElements++;
+        resultList.add(key);
+        resultList.add(value);
       }
+    }
 
-      if (numElements == count) {
-        break;
-      }
+    Integer numberOfIterationsCompleted = indexOfKeys - startCursor;
+
+    Pair resultsPair = new ImmutablePair(numberOfIterationsCompleted, resultList);
+
+    return resultsPair;
+  }
+
+  private int getCursorValueToReturn(int startCursor,
+      int numberOfIterationsCompleted,
+      List<ByteArrayWrapper> keySnapshot) {
+
+    if (startCursor + numberOfIterationsCompleted >= keySnapshot.size()) {
+      return 0;
     }
 
-    Pair<BigInteger, List<Object>> scanResult;
-    if (i >= size - 1) {
-      scanResult = new ImmutablePair<>(new BigInteger("0"), returnList);
-    } else {
-      scanResult = new ImmutablePair<>(new BigInteger(String.valueOf(i + 1)), returnList);
+    return (startCursor + numberOfIterationsCompleted);
+  }
+
+  @SuppressWarnings("unchecked")
+  private List<ByteArrayWrapper> getSnapShotOfKeySet(UUID clientID) {
+    List<ByteArrayWrapper> keySnapShot = this.hScanSnapShots.get(clientID);
+
+    if (keySnapShot == null) {
+      if (this.hScanSnapShots.isEmpty()) {
+        startHscanSnapshotScheduledRemoval();
+      }
+      keySnapShot = createKeySnapShot(clientID);
     }
-    return scanResult;
+    return keySnapShot;
+  }
+
+
+  private List<ByteArrayWrapper> createKeySnapShot(UUID clientID) {
+    List<ByteArrayWrapper> keySnapShot = new ArrayList<>();
+    keySnapShot.addAll(hash.keySet());
+
+    this.hScanSnapShots.put(clientID, keySnapShot);
+    this.hScanSnapShotCreationTimes.put(clientID, currentTimeMillis());
+
+    return keySnapShot;
   }
 
+
```

Comment:
please put back the formatting here through line 397, as well as your change to the formatting on lines 432&433

Reply:
not sure where "here" that you're referring to is?

Reply:
think I got what you were talking about.  Let me know if not